code/datums/components/anti_magic.dm

```diff
@@ -1,54 +1,120 @@
+/// This provides different types of magic resistance on an object
 /datum/component/anti_magic
-	var/magic = FALSE
-	var/holy = FALSE
-	var/psychic = FALSE
-	var/allowed_slots = ~ITEM_SLOT_BACKPACK
-	var/charges = INFINITY
-	var/blocks_self = TRUE
-	var/datum/callback/reaction
-	var/datum/callback/expire
-
-/datum/component/anti_magic/Initialize(_magic = FALSE, _holy = FALSE, _psychic = FALSE, _allowed_slots, _charges, _blocks_self = TRUE, datum/callback/_reaction, datum/callback/_expire)
+	var/antimagic_flags
+	var/charges
+	var/inventory_flags
+	var/datum/callback/drain_antimagic
+	var/datum/callback/expiration
+
+/**
+ * Adds magic resistances to an object
+ *
+ * Magic resistance will prevent magic from affecting the user if it has the correct resistance
+ * against the type of magic being used
+ * 
+ * args:
+ * * antimagic_flags (optional) A bitflag with the types of magic resistance on the object
+ * * charges (optional) The amount of times the object can protect the user from magic 
+ * * inventory_flags (optional) The inventory slot the object must be located at in order to activate
+ * * drain_antimagic (optional) The proc that is triggered when an object has been drained a antimagic charge
+ * * expiration (optional) The proc that is triggered when the object is depleted of charges
+ * *
+ * antimagic bitflags: (see code/__DEFINES/magic.dm)
+ * * MAGIC_RESISTANCE - Default magic resistance that blocks normal magic (wizard, spells, staffs)
+ * * MAGIC_RESISTANCE_MIND - Tinfoil hat magic resistance that blocks mental magic (telepathy, abductors, jelly people)
+ * * MAGIC_RESISTANCE_HOLY - Holy magic resistance that blocks unholy magic (revenant, cult, vampire, voice of god)
+ * * MAGIC_RESISTANCE_ALL - All magic resistances combined
+**/
+/datum/component/anti_magic/Initialize(
+		antimagic_flags = MAGIC_RESISTANCE,
+		charges = INFINITY, 
+		inventory_flags = ~ITEM_SLOT_BACKPACK, // items in a backpack won't activate, anywhere else is fine
+		drain_antimagic, 
+		expiration
```

Comment:
```suggestion
		datum/callback/drain_antimagic, 
		datum/callback/expiration
```

Reply:
Also, these callbacks need to be cleaned up in `Destroy()` if i'm not mistaken.

Reply:
https://github.com/tgstation/tgstation/pull/64124/commits/e02f6670519cd7fe7eb90c3a7b3efed07254e458 I'm guessing this causes hard deletes if they aren't properly cleaned up?